crates/eips/src/eip7702.rs

```diff
@@ -0,0 +1,158 @@
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+#[cfg_attr(
+    any(test, feature = "arbitrary"),
+    derive(proptest_derive::Arbitrary, arbitrary::Arbitrary)
+)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+pub struct Authorization {
+    /// Chain ID
+    pub chain_id: ChainId,
+    /// The address of the code that will get set to the signer's address
+    pub address: Address,
+    /// Optional nonce
+    pub nonce: Option<u64>,
+    /// yParity: Signature Y parity
+    pub y_parity: bool,
+    /// The R field of the signature
+    pub r: U256,
+    /// The S field of the signature
+    pub s: U256,
+}
+
+impl Authorization {
+    fn fields_length(&self) -> usize {
+        let mut length = 0;
+        length += self.chain_id.length();
+        length += self.address.length();
+        length += self.nonce.map(|n| vec![n]).unwrap_or(vec![]).length();
+        length += self.y_parity.length();
+        length += self.r.length();
+        length += self.s.length();
+        length
+    }
+
+    /// Calculates a heuristic for the in-memory size of the [`Authorization`]
+    #[inline]
+    pub fn size(&self) -> usize {
+        mem::size_of::<ChainId>()
+            + mem::size_of::<Address>()
+            + mem::size_of::<Option<u64>>()
+            + mem::size_of::<bool>()
+            + mem::size_of::<U256>()
+            + mem::size_of::<U256>()
+    }
+}
+
+impl Encodable for Authorization {
+    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
+        let list_header = alloy_rlp::Header { list: true, payload_length: self.fields_length() };
+        list_header.encode(out);
+        self.chain_id.encode(out);
+        self.address.encode(out);
+        self.nonce.map(|n| vec![n]).unwrap_or(vec![]).encode(out);
+        self.y_parity.encode(out);
+        self.r.encode(out);
+        self.s.encode(out);
+    }
+}
+
+impl Decodable for Authorization {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let rlp_head = alloy_rlp::Header::decode(buf)?;
+        if !rlp_head.list {
+            return Err(alloy_rlp::Error::UnexpectedString);
+        }
+
+        let started_len = buf.len();
+        let chain_id: ChainId = Decodable::decode(buf)?;
+        let address: Address = Decodable::decode(buf)?;
+        let nonce_list: Vec<u64> = Decodable::decode(buf)?;
+        let nonce = nonce_list.first().copied();
+        let y_parity = Decodable::decode(buf)?;
+        let r = Decodable::decode(buf)?;
+        let s = Decodable::decode(buf)?;
+
+        let consumed = started_len - buf.len();
+        if consumed != rlp_head.payload_length {
+            return Err(alloy_rlp::Error::ListLengthMismatch {
+                expected: rlp_head.payload_length,
+                got: consumed,
+            });
+        }
+        Ok(Self { chain_id, address, nonce, y_parity, r, s })
+    }
+}
```

Comment:
For encoding it might be easier to create another struct, where the nonce is in a vec, and derive encodable / decodable on that struct, with conversions to/from Authorization instead of doing a manual implementation.